// fed/libs/fed_imgen/index.js
var express, fs, imgGen, os, path;

express = require("express");

imgGen = require("./imageGen");

fs = require("fs");

os = require("os");

path = require("path");

exports.init = function(config) {
  var cfg;
  cfg = {
    cacheTimeout: config.timeout || 5000,
    url: config.url || "/_",
    defaultColor: [0xff, 0x01, 0xa2, 0.5],
    destPath: config.destPath || os.tmpDir()
  };
  return this.on("appinit2", function(app) {
    return app.get("" + cfg.url + "/:imageParams", function(req, res, next) {
      var color, colorParam, fileName, height, imageParams, size, width, _ref, _ref1, _ref2;
      imageParams = req.param("imageParams");
      _ref = imageParams.split(/-/), size = _ref[0], color = _ref[1];
      _ref1 = size.split(/x|X/), width = _ref1[0], height = _ref1[1];
      if (!height) {
        height = width;
      }
      _ref2 = [parseInt(width), parseInt(height)], width = _ref2[0], height = _ref2[1];
      if (!width) {
        next();
        return;
      }
      colorParam = color ? color.split(",") : cfg.defaultColor;
      colorParam = [parseFloat(colorParam[0] || cfg.defaultColor[0]), parseFloat(colorParam[1] || cfg.defaultColor[1]), parseFloat(colorParam[2] || cfg.defaultColor[2]), parseFloat(colorParam[3] || cfg.defaultColor[3])];
      fileName = path.join(cfg.destPath, "" + width + "x" + height + "-" + (colorParam.join('_')) + ".png");
      if (!fs.existsSync(fileName)) {
        imgGen.createImage(fileName, width, height, colorParam);
        setTimeout(function() {
          return fs.unlinkSync(fileName);
        }, cfg.cacheTimeout);
      }
      res.sendfile(fileName);
    });
  });
};
